File: type2FLS.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

class Type2FuzzySystem {
public:
    Type2FuzzySystem() {
        // Define temperature ranges for fuzzy sets
        tempRanges["Cold"] = {0, 10, 20};
        tempRanges["Neutral"] = {15, 25, 35};
        tempRanges["Hot"] = {30, 40, 50};
        
        // Define Footprint of Uncertainty (FOU) for Type-2 Fuzzy Sets
        fou["Cold"] = {0.2, 0.3};
        fou["Neutral"] = {0.3, 0.4};
        fou["Hot"] = {0.2, 0.3};
        
        // Define fan speed fuzzy sets
        fanSpeedRanges["Slow"] = {0, 25, 50};
        fanSpeedRanges["Medium"] = {25, 50, 75};
        fanSpeedRanges["Fast"] = {50, 75, 100};
    }

    // Membership function for temperature fuzzy sets
    double membershipFunction(double x, const std::string& fuzzySet) {
        const auto& range = tempRanges[fuzzySet];
        double a = range[0], b = range[1], c = range[2];

        if (x <= a || x >= c) return 0;
        if (x > a && x < b) return (x - a) / (b - a);
        if (x >= b && x < c) return (c - x) / (c - b);
        return 0;
    }

    // Compute fuzzy set values considering FOU
    std::vector<double> fuzzySet(double x, const std::string& fuzzySet) {
        double primaryMembership = membershipFunction(x, fuzzySet);
        double fMin = fou[fuzzySet][0];
        double fMax = fou[fuzzySet][1];

        return {std::max(0.0, std::min(primaryMembership + fMax, 1.0)),
                std::max(0.0, std::min(primaryMembership - fMin, 1.0))};
    }

    // Apply fuzzy rules and compute fan speed
    std::vector<double> applyRules(double temp) {
        std::vector<double> fanSpeeds = {0, 0, 0}; // Slow, Medium, Fast

        auto coldFuzzy = fuzzySet(temp, "Cold");
        auto neutralFuzzy = fuzzySet(temp, "Neutral");
        auto hotFuzzy = fuzzySet(temp, "Hot");

        if (coldFuzzy[0] > 0) fanSpeeds[0] += coldFuzzy[0]; // Slow
        if (neutralFuzzy[0] > 0) fanSpeeds[1] += neutralFuzzy[0]; // Medium
        if (hotFuzzy[0] > 0) fanSpeeds[2] += hotFuzzy[0]; // Fast

        return fanSpeeds;
    }

    // Defuzzification
    double defuzzify(const std::vector<double>& fanSpeeds) {
        double weightedSum = 0;
        double totalWeight = 0;

        auto addToSum = [&](const std::string& speed, double value) {
            if (value > 0) {
                auto midPoint = (fanSpeedRanges[speed][0] + fanSpeedRanges[speed][2]) / 2.0;
                weightedSum += midPoint * value;
                totalWeight += value;
            }
        };

        addToSum("Slow", fanSpeeds[0]);
        addToSum("Medium", fanSpeeds[1]);
        addToSum("Fast", fanSpeeds[2]);

        return totalWeight > 0 ? weightedSum / totalWeight : 0;
    }

    // Run the fuzzy system
    double run(double temperature) {
        auto fanSpeeds = applyRules(temperature);
        return defuzzify(fanSpeeds);
    }

private:
    std::map<std::string, std::vector<double>> tempRanges;
    std::map<std::string, std::vector<double>> fou;
    std::map<std::string, std::vector<double>> fanSpeedRanges;
};

int main() {
    Type2FuzzySystem fuzzySystem;

    // Define temperature input
    double temperatureInput = 28;

    // Get the fan speed output
    double fanSpeed = fuzzySystem.run(temperatureInput);
    std::cout << "Recommended Fan Speed: " << fanSpeed << std::endl;

    return 0;
}

File: type2FLS.py
import numpy as np
import matplotlib.pyplot as plt

class Type2FuzzySystem:
    def __init__(self):
        # Define the temperature ranges for fuzzy sets
        self.temp_ranges = {
            'Cold': [0, 10, 20],
            'Neutral': [15, 25, 35],
            'Hot': [30, 40, 50]
        }
        
        # Define the fan speed fuzzy sets
        self.fan_speed_ranges = {
            'Slow': [0, 25, 50],
            'Medium': [25, 50, 75],
            'Fast': [50, 75, 100]
        }
        
        # Define Footprint of Uncertainty (FOU) for Type-2 Fuzzy Sets
        self.fou = {
            'Cold': [0.2, 0.3],
            'Neutral': [0.3, 0.4],
            'Hot': [0.2, 0.3]
        }
    
    def membership_function(self, x, fuzzy_set):
        a, b, c = self.temp_ranges[fuzzy_set]
        if x <= a or x >= c:
            return 0
        elif a < x < b:
            return (x - a) / (b - a)
        elif b <= x < c:
            return (c - x) / (c - b)
    
    def fuzzy_set(self, x, fuzzy_set):
        primary_membership = self.membership_function(x, fuzzy_set)
        f_min, f_max = self.fou[fuzzy_set]
        return [max(0, min(primary_membership + f_max, 1)), max(0, min(primary_membership - f_min, 1))]
    
    def apply_rules(self, temp):
        fan_speeds = {'Slow': 0, 'Medium': 0, 'Fast': 0}
        
        cold_fuzzy = self.fuzzy_set(temp, 'Cold')
        neutral_fuzzy = self.fuzzy_set(temp, 'Neutral')
        hot_fuzzy = self.fuzzy_set(temp, 'Hot')
        
        if cold_fuzzy[0] > 0:
            fan_speeds['Slow'] += cold_fuzzy[0]
        
        if neutral_fuzzy[0] > 0:
            fan_speeds['Medium'] += neutral_fuzzy[0]
        
        if hot_fuzzy[0] > 0:
            fan_speeds['Fast'] += hot_fuzzy[0]
        
        return fan_speeds

    def defuzzify(self, fan_speeds):
        weighted_sum = 0
        total_weight = 0
        for speed, value in fan_speeds.items():
            if value > 0:
                midpoint = np.mean(self.fan_speed_ranges[speed])
                weighted_sum += midpoint * value
                total_weight += value
        return weighted_sum / total_weight if total_weight > 0 else 0

    def run(self, temperature):
        fan_speeds = self.apply_rules(temperature)
        return self.defuzzify(fan_speeds)

# Example Usage
fuzzy_system = Type2FuzzySystem()

# Define temperature input
temperature_input = 28

# Get the fan speed output
fan_speed = fuzzy_system.run(temperature_input)
print(f"Recommended Fan Speed: {fan_speed:.2f}")

# Plot the fuzzy membership functions and the result
x = np.linspace(0, 50, 100)
y_cold = [fuzzy_system.membership_function(val, 'Cold') for val in x]
y_neutral = [fuzzy_system.membership_function(val, 'Neutral') for val in x]
y_hot = [fuzzy_system.membership_function(val, 'Hot') for val in x]

plt.plot(x, y_cold, label='Cold')
plt.plot(x, y_neutral, label='Neutral')
plt.plot(x, y_hot, label='Hot')
plt.xlabel('Temperature')
plt.ylabel('Membership Degree')
plt.title('Temperature Fuzzy Sets')
plt.legend()
plt.grid(True)
plt.show()
